### PR TITLE
3rd Edition Questions

### PR DESCRIPTION
**Pull request by rajarshimaitra**

### CODE REVIEW

**Comment by QureshiFaisal** [comment]
I will curate question banks for chapter 11 The Blockchain  & chapter 12 Mining and Consensus

**Comment by theanmolsharma** [comment]
I'll take up chapters 4 to 9. I am tracking progress in this comment.

- [x] Chapter 4 - Keys and Addresses
- [x] Chapter 5 - Wallet Recovery
- [x] Chapter 6 - Transactions
- [x] Chapter 7 - Authorization and Authentication
- [x] Chapter 8 - Digital Signatures
- [x] Chapter 9 - Transaction Fees

**Comment by setusaurabh** [comment]
For session 1 i.e chapter 1 & 2. 

Meeting the earlier prepared questions and assembled it into these set of 16 questions, as i needed to send today.

Round 1:

1. What is the difference between the Bitcoin network and Bitcoin as a currency? Why doesn't Bitcoin have a hierarchical network?
2. What is the double-spend problem and how did Bitcoin solve it? What determines the rate of inflation in Bitcoin?
3. What kind of privacy is there for transactions on a public ledger? Do you think it would be better if Bitcoin transactions were reversible?
4. Do you believe that Bitcoin needs to be competitive with Visa/Mastercard to succeed? How does Bitcoin's decentralization contribute to its security and resilience?
5. What's the smallest unit of Bitcoin? What does the receiver need to provide to receive Bitcoin from the sender?
6. What basic information do you need to construct a transaction? Why are transactions called "chains of digital signatures"? Explain spending a coin as extending the "chain of digital signatures" one step further.
7. What is the status of a transaction that has not yet been included in a block but has been propagated around the network? What is a transaction fee? Why is it required? How is it calculated?
8. What is change? When is change needed? When should change be included, and when should it not be included?

Round 2:

9. What is Proof of Work mining, and how is it like a needle-in-a-haystack problem of random draws? What is a mining pool, and why would someone belong to one?
10. What is a reorg (reorganization)? How does the network decide which of two blocks to add to the chain when they are at the same height?
11. What is difficulty adjustment? Why do some services require six confirmations to clear funds?
12. How can Bitcoin transactions be constructed and broadcast? What mechanisms ensure their propagation across the network?
13. How does Bitcoin ensure the validity and integrity of transactions through cryptographic signatures?
14. Why is the concept of "change" necessary in Bitcoin transactions? How does it affect the creation of transactions?
15. What is the role of mining pools in the Bitcoin network? How do they influence the distribution of mining rewards?
16. How does the network handle competing blocks at the same height? What factors influence the selection of the canonical chain?

**Comment by QureshiFaisal** [comment]
> I'll take up chapters 4 to 9. I am tracking progress in this comment.
> 
> * [ ]  Chapter 4 - Keys and Addresses
> * [ ]  Chapter 5 - Wallet Recovery
> * [ ]  Chapter 6 - Transactions
> * [ ]  Chapter 7 - Authorization and Authentication
> * [ ]  Chapter 8 - Digital Signatures
> * [ ]  Chapter 9 - Transaction Fees

I'd be happy to take up chapters 6 and 7 if you're occupied with other commitments. Kindly let me know.

**Comment by QureshiFaisal** [comment]
@rajarshimaitra I added questions for chapter 11. Kindly proof-read and approve. Will then proceed with remaining chapters. I would like to take up 10-14.

File: questions-3rd-ed.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+--------------
+
+# Week 6
+
+## Chapter 11: The Blockchain
+
+- How is the Bitcoin blockchain analogous to layers in a geological formation?
+- Describe the structure of a block and what data does a block header contain?
+- What are the two identifiers of a Block? Discuss each in brief.
+- What is a Merkle Tree and what significance does it have in Bitcoin?
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Can we reframe something like, "What is a Markle Tree and what is it used for in the blockchain?" "Implication in Bitcoin" is a bit ambiguous.

File: questions-3rd-ed.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+- What is a design flaw in the Merkle Tree used in Bitcoin?
+- What is a Light Weight Client in context to Bitcoin? How do Light Weight Clients verify transactions associated to their addresses?
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Aren't they called just Light Clients?

File: questions-3rd-ed.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+- Explain mainnet and testnet in brief.
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: A bit more explicite like

Whats the difference between mainnet and testnet? Can testnet coins have value?

File: questions-3rd-ed.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+- Explain signet and regtest in brief.
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Similarly,

Whats the difference between Signet and Regtest? When would you use regtest for testing, and when Signet?

File: questions-3rd-ed.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+  
+## Chapter 12: Mining and Consensus
+
+- What are the goals of Bitcoin mining and what are it’s incentives?
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Do you mean the incentive for miners to mine bitcoin? can be a bit more explicit.

File: questions-3rd-ed.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+- Explain in brief how does the Bitcoin network attain emergent decentralised consensus.
+- What is a coin-base transaction? What prevents the miners from inflating the coin-base reward?
+- What is Mining Difficulty? How is it adjusted?
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Can we also add, Why is it adjusted?

File: questions-3rd-ed.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+- Explain the 51% attack. Can it compromise the security of private keys and signing algorithms?
+- What factors could cause a change in Bitcoin’s consensus rules? How is a change in consensus rules achieved in a decentralised network like Bitcoin?
+- Explain a Hard fork. Is a new software implementation of the consensus rules a pre-requisite for it?
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md]: Can it be clarified a bit more? Also better to avoid questions, which can be answered with a Yes/No.

**Review by rajarshimaitra — COMMENTED** [review]
@QureshiFaisal left some comments for you.

File: questions-3rd-ed.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+## Chapter 4: Keys and Addresses
+
+### Elliptic Curves
+
+- Describe the functions of ‘private keys’, ‘public keys’, ‘Bitcoin addresses’, and ‘digital signatures’. What is the mathematical relationship between them and how are they used in Bitcoin?
+
+- What is the discrete logarithm problem, and why is it considered hard to solve?
+
+- What is asymmetry in cryptography, and why are one-way functions useful? Identify a one-way cryptographic function used in deriving public keys from private keys.
```
Comment (rajarshimaitra) [inline questions-3rd-ed.md:53]: "identify the one-way problem"? There is only one right for priv > pub calculation?

**Review by rajarshimaitra — COMMENTED** [review]
@theanmolsharma Chap 4 looks good. Just one comment.

**Comment by rajarshimaitra** [comment]
@setusaurabh Below are the list of the questions for Chap 3 and 4. 

We have more Qs than we probably need. So feel free to select among them, as many as you need to adjust for the group participant count. 

We can post the remaining unselected ones in a `Bonus Question` section in the instruction sheet.

## Chapter 3: Bitcoin Core

- Why is Bitcoin Core called a "Reference Implementation"? What other implementations of Bitcoin are out there? Is it preferable to have many implementations of Bitcoin?
- What are BIPs (Bitcoin Improvement Proposals)? What are their role in the Bitcoin development process?
- What are the few major configuration options in Bitcoin Core? As a home node operator, which one do you think is most useful?
- What is `txindex`? When would a node operator use this configuration flag? What are the overheads of turning this configuration on?
- What is `bitcoin-cli`? What is it used for? Have you explored any `bitcoin-cli` commands? Which commands do you find most useful as a home node operator?
- What is the approximate disk space requirement for running Bitcoin Core? What are the approximate network bandwidth and memory requirements? Do you think it's "cheap" to run a node?
- What are commands you would use sequentially to get the fully deserialized coinbase transaction of the Block 100, using `bitcoin-cli`?
- How can you verify the signatures of download binaries of Bitcoin Core? What are the potential damages of running a malicious version of the software?


## Chapter 4: Keys and Addresses

- Describe the functions of ‘private keys’, ‘public keys’, ‘Bitcoin addresses’, and ‘digital signatures’. What is the mathematical relationship between them and how are they used in Bitcoin?

- What is the discrete logarithm problem, and why is it considered hard to solve?

- What is asymmetry in cryptography, and why are one-way functions useful? Identify a one-way cryptographic function used in deriving public keys from private keys.

- Describe instances where private keys have been compromised due to lack of entropy. Why is it important to use a cryptographically secure pseudorandom number generator to produce a private key? What are some techniques of generating pseudo-random numbers?

- What is secp256k1, and what are the curve constants used in Bitcoin? How is ECDSA malleable?

- What is a Bitcoin address, and why are addresses used instead of raw public keys?

- What hash functions are used in the generation of Bitcoin addresses?

- Why is Base58 used for Bitcoin addresses instead of Base64? Would you make any modifications beyond Base58?

- What are the different private key formats used by wallets?

- What are compressed public keys, and what are the benefits of compressing a public key?

- What are the different kinds of Bitcoin addresses? Explain the differences between P2PKH and P2SH.

- What is the difference between a hot wallet and cold storage?

- What is bech32? How is it better than base58? Are there any problems in bech32? How are they solved?

File: questions-3rd-ed.md (hunk 8 above)
Comment (theanmolsharma) [reply questions-3rd-ed.md:53]: yes, it's `A = a.G`

**Comment by QureshiFaisal** [comment]
@rajarshimaitra have made changes suggested by you and have added questions for chapter 14.

**Review by rajarshimaitra — COMMENTED** [review]
ACK

**Pull request merged**